internal/certdb/certdb.go

```diff
@@ -0,0 +1,136 @@
+	ID          int
+	CSR         string
+	Certificate string
+}
+
+// RetrieveAll gets every CertificateRequest entry in the table.
+func (db *CertificateRequestsRepository) RetrieveAll() ([]CertificateRequest, error) {
+	rows, err := db.conn.Query(fmt.Sprintf(queryGetAllCSRs, db.table))
+	if err != nil {
+		return nil, err
+	}
+
+	var allCsrs []CertificateRequest
+	defer rows.Close()
+	for rows.Next() {
+		var csr CertificateRequest
+		if err := rows.Scan(&csr.ID, &csr.CSR, &csr.Certificate); err != nil {
+			return nil, err
+		}
+		allCsrs = append(allCsrs, csr)
+	}
+	return allCsrs, nil
+}
+
+// Retrieve gets a given CSR from the repository.
+// It returns the row id and matching certificate alongside the CSR in a CertificateRequest object.
+func (db *CertificateRequestsRepository) Retrieve(csr string) (CertificateRequest, error) {
+	var newCSR CertificateRequest
+	row := db.conn.QueryRow(fmt.Sprintf(queryGetCSR, db.table), csr)
+	if err := row.Scan(&newCSR.ID, &newCSR.CSR, &newCSR.Certificate); err != nil {
+		newCSR.ID = -1
+		newCSR.Certificate = ""
+		newCSR.CSR = ""
+		return newCSR, err
+	}
+	return newCSR, nil
+}
+
+// Create creates a new entry in the repository.
+// The given CSR must be valid and unique
+func (db *CertificateRequestsRepository) Create(csr string) (int64, error) {
+	if err := ValidateCertificateRequest(csr); err != nil {
+		return 0, err
+	}
+	result, err := db.conn.Exec(fmt.Sprintf(queryCreateCSR, db.table), csr)
+	if err != nil {
+		return 0, err
+	}
+	id, err := result.LastInsertId()
+	if err != nil {
+		return 0, err
+	}
+	return id, nil
+}
+
+// Update adds a new cert to the given CSR in the repository.
+// The given certificate must share the public key of the CSR and must be valid.
+func (db *CertificateRequestsRepository) Update(csr string, cert string) (int64, error) {
+	if err := ValidateCertificate(cert, csr); err != nil {
+		return 0, err
+	}
+	result, err := db.conn.Exec(fmt.Sprintf(queryUpdateCSR, db.table), cert, csr)
+	if err != nil {
+		return 0, err
+	}
+	id, err := result.LastInsertId()
+	if err != nil {
+		return 0, err
+	}
+	return id, nil
+}
+
+// Delete removes a CSR from the database alongside the certificate that may have been generated for it.
+func (db *CertificateRequestsRepository) Delete(csr string) error {
+	_, err := db.conn.Exec(fmt.Sprintf(queryDeleteCSR, db.table), csr)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+// Close closes the connection to the repository cleanly.
+func (db *CertificateRequestsRepository) Close() error {
+	if db.conn == nil {
+		return nil
+	}
+	if err := db.conn.Close(); err != nil {
+		return err
+	}
+	return nil
+}
+
+// NewCertificateRequestsRepository connects to a given table in a given database,
+// stores the connection information and returns an object containing the information.
+// The database path must be a valid file path or ":memory:".
+// The table will be created if it doesn't exist in the format expected by the package.
```

Comment:
Is there a way to enforce docstring in public functions? We are already using `golangci-lint`, maybe we can simply enable one of its linters. In any case, this can be enabled in a PR of its own.

Reference:
- https://golangci-lint.run/usage/linters/

Reply:
I read the disabled ones but none of them stuck out for me. We could eventually write one ourselves.